design/design-doc-agent.md

```diff
@@ -0,0 +1,586 @@
+users want to manage from a single central point is infrastructure (including
+policy, audit etc) but not necessarily applications.
+
+## Proposal
+
+In order to preserve the central infrastructure management ability while
+alleviating the issues above, we will change our approach from push-based one to
+a pull-based one where applications, and their requirements are deployed into the
+remote cluster, and they request the infrastructure from central cluster and pull
+the necessary credentials.
+
+Since having this logic in the applications themselves wouldn't be a good UX, we
+will have an agent that you will need to deploy into your remote cluster for
+doing the heavy-lifting for you. There are several technical problems to be
+solved in order to make the experience smooth. Overall, the goal is that we want
+to keep the UX of local mode for application operators while keeping the power
+of centralized infrastructure management for platform operators. For reference,
+here is an example local mode experience we'd like to have for the remote mode as
+well:
+
+```yaml
+apiVersion: common.crossplane.io/v1alpha1
+kind: MySQLInstanceRequirement
+metadata:
+  name: sqldb
+  namespace: default
+spec:
+  version: "5.7"
+  storageGB: 20
+  writeConnectionSecretToRef:
+    name: sql-creds
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: wp
+  namespace: default
+spec:
+containers:
+- name: wordpress
+    image: "wordpress:4.6.1-apache"
+    env:
+    - name: WORDPRESS_DB_PASSWORD
+    valueFrom:
+        secretKeyRef:
+          name: sql-creds
+          key: password
+```
+
+The agent will be a Kubernetes controller running in the remote cluster and
+watching all `*Requirement` types. Next sections will talk about the
+implementation and user experience we'd like to have.
+
+### Synchronization
+
+In local mode, users directly interact with what's published by the platform
+team, which is `*Requirement` types and consume the infrastructure by mounting
+the secret whose name they specified on the `*Requirement` custom resource. To
+keep this experience, we need to have a synchronization loop for the following
+resources:
+
+* Pull
+  * `CustomResourceDefinition`s of all types that we want the applications to be
+    able to manage and view:
+    * Requirements that are published via `InfrastructurePublication`s. The
+      source of truth will be the remote cluster.
+    * `InfrastructureDefinition`, `InfrastructurePublication` and `Composition`.
+  * `Composition`s to discover what's available to choose. CRs of this type will
+    be read-only, and the source of truth will be the central cluster.
+  * `InfrastructurePublication`s to discover what's availabe as published API.
+    Read-only.
+  * `InfrastructureDefinition`s to discover how the secret keys are shaped.
+    Read-only.
+  * `Secret`s that are result of the infrastructure that is provisioned so that
+    it can be mounted to `Pod`s. Read-only.
```

Comment:
> I imagine we could pull only the IDs and IPs, and have the agent create the appropriate CRDs in the remote cluster.

I wouldn't really bind agent to the generation logic as it'd imply a manually enforced parity between the generation logic here and in the Crossplane itself. I think just copying the CRDs of the resources listed in the first bullet would be sufficient and it'd guarantee that the source of truth is one and it's Crossplane in the central cluster.

Reply:
That makes sense, but I do think we should create a record (i.e. IP) on the remote cluster so that folks can have the same `kubectl get infrastructurepublications` experience on remote clusters as they can on the central cluster.

Reply:
> but I do think we should create a record (i.e. IP) on the remote cluster so that folks can have the same kubectl get infrastructurepublications experience on remote clusters as they can on the central cluster.

Yeah, that's the current proposal. I'll try to make the wording more clear. All CRDs as well as IPs and IDs will be copied, it's just there won't be any controllers watching IP/IDs, their corresponding CRDs will be brought over by the agent.